File: reynoldsSim.R

# THE FOLLOWING IS A 2-D VARIANT OF CRAIG REYNOLDS' BOID MODEL.
# Written by Gabriel Dell'Accio
# Algorithm inspired by Conrad Parker's Boid's pseudocode at: 
# http://www.kfish.org/boids/pseudocode.html

# Overview:
#   Rules for swarming
#     1) Center of Mass: Boids are attracted to the center of the overall swarm.
#     2) Avoidance rule: Boids avoid collision with other boids by avoiding coming within
#        a certain distance of one another.
#     3) Match Velocity: Boids can sense the average velocity of the swarm and do their 
#        best to match it.

# LIBRARIES:
library(ggplot2)

# ARENA PARAMETERS:
  # Variable to hold the x,y center coordinate of the arena. 
  xLength <- 2000.0
  yLength <- 2000.0

  centerCor <- rep(xLength/2,2)
# SWARM PARAMETERS:
  # number of individuals
  nIndividuals <- 100 
  swarmCohesion <- 0.5
  swarmAvoidance <- 0.5
  baseVelocity <- 20
  densitySensitivity <- 75
  spatialDistribution <- 100
  thetaDistribution <- 50
  step.iterations <- 200

  # velocityList <- vector("list", nIndividuals)
  # print(velocityList)

  xVelocity <- rep(10, nIndividuals)
  yVelocity <- rep(10, nIndividuals)

  # This function creates the arena.
  create.arena <- function(xLength, yLength) {
    
    # length of x plane 
    xLimit <- xLength 
    # Length of y plane
    yLimit <- yLength
    
    # temporary data frame that holds the data 
    # that creates the borders of the 2D arena. 
    arena.Data <<- data.frame(
      yWall = rep(1,yLimit),
      xWall = c(1:xLimit),
      xNorthWall = c(1:xLimit),
      yNorthWall = rep(yLimit,yLimit)
    )

    # Plot Arena Borders
    arena <- ggplot(data = arena.Data, aes(x=xWall,y=yWall))
    arena <- arena + geom_point() +  ylim(1,yLimit) + xlim(1,xLimit) + geom_point(aes(xNorthWall,yNorthWall)) + geom_point(aes(rep(1,xLimit),c(1:yLimit))) + geom_point(aes(rep(xLimit,xLimit),c(1:yLimit)))

    return(arena)
  }
  arenaSim <- create.arena(xLength, yLength)

  # This function helps initialize the swarm and arena using a data frame, arena.data, 
  # that stores information about the swarm and the arena. 
  initialize.swarm <- function(avoidance, spatial.distribution.factor,coordinateLocation) {
      arena.Data$index <<- c(1:nIndividuals)
      arena.Data$xPosition <<- rep(NA, xLength)
      arena.Data$yPosition <<- rep(NA, yLength)
      # arena.Data$theta <<- rep(NA, xLength)
      # velocityList <<- rep(baseVelocity, nIndividuals)
    
    # Create the x,y position of each individual in the swarm by drawing from a 
    # random normal distribution around a given coordinate location in the arena. 
    for(var in 1:nIndividuals) {
      # velocityList[] <- c(baseVelocity,baseVelocity)
      arena.Data$xPosition[var] <<- rnorm(1, mean=coordinateLocation[1], sd=spatial.distribution.factor + 50)
      arena.Data$yPosition[var] <<- rnorm(1, mean=coordinateLocation[2], sd=spatial.distribution.factor + 50)
      # arena.Data$theta[var] <<- rnorm(1, mean = 180, sd= theta.distribution.factor + 5)
    }
  }

  # Call the initialize function that creates teh arena.data data frame 
  # with all previously declared swarm and arena parameters.
  initialize.swarm(swarmAvoidance, spatialDistribution, centerCor)

  # Function that updates the current arena with the swarm's current location.
  display.swarm <- function() {
    return(arenaSim + geom_point(aes(arena.Data$xPosition, arena.Data$yPosition)))
  }

arenaSim <- display.swarm()
arenaSim

# GSI: functions for computing the group stability index
  # This function computes the distance between two x,y points in the 2D arena. 
  compute.cart.distance  <- function(x1,x2,y1,y2) {
    return(sqrt(((x1-x2)^2) + ((y1-y2)^2)) )
  }
  # This function calculates the total distance between all members of the flock. It returns the total distance between all individuals in one step of the arena.
  calculate.group.distance <- function() {
    groupedDistance <- 0
    for(var in 1:nIndividuals) {
      for(varTwo in 1:nIndividuals) {
         dist <- compute.cart.distance(arena.Data$xPosition[var], arena.Data$xPosition[varTwo], arena.Data$yPosition[var], arena.Data$yPosition[varTwo])
         groupedDistance <- groupedDistance + dist
      }
    }
    return(groupedDistance)
  }
  # This function takes in the previous time step's total group distance and the current time steps group distance to calculate the 
  # group stability over time.
  calculate.gsi <- function(groupedDistancePrev, groupedDistanceNext) {
    finalGsi <- 1 - (((abs(groupedDistancePrev - groupedDistanceNext)) / 4 ) / (nIndividuals*(nIndividuals-1)/2))
    return(finalGsi)
  }


# A function that returns the average x,y coordinate of the swarm.
find.center <- function() {

  totalX <- 0
  totalY <- 0


  for(var in 1:nIndividuals) {
    totalX <- totalX + arena.Data$xPosition[var] 
    totalY <- totalY + arena.Data$yPosition[var] 
  }

  averageX <- totalX / nIndividuals
  averageY <- totalY / nIndividuals

  centerMass <- c(averageX, averageY)

  return(centerMass)

}


# RULE 1: calculate the swarm's center of mass 
center.of.mass <- function(boid, count) {

  totalX <- 0
  totalY <- 0


  for(var in 1:nIndividuals) {
    totalX <- totalX + arena.Data$xPosition[var] 
    totalY <- totalY + arena.Data$yPosition[var] 
  }

  averageX <- totalX / nIndividuals
  averageY <- totalY / nIndividuals

  centerMass <- c((averageX - boid[1]/100 ), (averageY - boid[2]/100))

  # print(paste("center of mass: ", centerMass))
  return(centerMass)

}


# RULE 2: calculate the avoidance correction for a boid   
avoidance.rule <- function(boid, count) {

  correctedCourse <- c(0,0)

  for(var in 1:nIndividuals) {

    if(count != var) {

      if((abs(arena.Data$xPosition[var] - boid[1]) < 100) && (abs(arena.Data$yPosition[var] - boid[2]) < 100)) {
        b.position <- c(arena.Data$xPosition[var], arena.Data$yPosition[var])

        correctedCourse <- correctedCourse - (b.position - boid)
      } 


    }
  }

  # print(paste("corrected avoidance: ", correctedCourse))
  return(correctedCourse)
}

# RULE 3: calculate the match velocity correction for a boid
match.velocity <- function(boid, count) {

  meanVelocityX <- 0 
  meanVelocityY <- 0

  for(var in 1:nIndividuals) {
    meanVelocityX <- meanVelocityX + xVelocity[var]
    meanVelocityY <- meanVelocityY + yVelocity[var]  
  }


  meanVelocityX <- meanVelocityX / nIndividuals
  meanVelocityY <- meanVelocityY / nIndividuals

  velocityCorrectionX <- (meanVelocityX - xVelocity[count]) /8
  velocityCorrectionY <- (meanVelocityY - xVelocity[count]) /8


  return(c(velocityCorrectionX, velocityCorrectionY))
}
 
# This function steps the swarm one step using the the 3 behavior rules 
# written above. The input to this function, "simVersion", is either "reynolds" 
# or "asocial". 
compute.next.pos <- function() {

  for(var in 1:nIndividuals) {

      currentIndividualLoc <- c(arena.Data$xPosition[var], arena.Data$yPosition[var])
      
      # Rule 1
      centerMassRule <- center.of.mass(currentIndividualLoc, var)
      # Rule 2 
      avoidanceRule <- avoidance.rule(currentIndividualLoc, var)
      # Rule 3
      velocityRule <- match.velocity(currentIndividualLoc, var)
      
      

      currentIndividualLoc <- currentIndividualLoc + c(xVelocity[var],yVelocity[var])

      xVelocity[var] <<- xVelocity[var] + centerMassRule[1] + avoidanceRule[1] + velocityRule[1]
      yVelocity[var] <<- yVelocity[var] + centerMassRule[2] + avoidanceRule[2] + velocityRule[2]


      arena.Data$xPosition[var] <<- currentIndividualLoc[1]
      arena.Data$yPosition[var] <<- currentIndividualLoc[2]
  }

}



# This function steps the swarm x iterations and tracks the group stability of the
# swarm in each time step. 
time.stepper <- function(iterations) {

  gsiLog <- rep(0,iterations)

  # For i iterations ...
  for(i in 1:iterations) {
    
    # calculate total group distance at this time step 
    prevTotalDistance <- calculate.group.distance()

    # Compute new positons for each individual in the swarm.
    compute.next.pos()

    # Calculate the total group distance after moving the swarm.
    nextTotalDistance <- calculate.group.distance()

    # Calculate the change in group stability between the previous and current 
    # time step and log it in the gsiLog vector.
    gsiLog[i] <- calculate.gsi(prevTotalDistance, nextTotalDistance)

    # Display the swarm.
    plot(arena.Data$xPosition, arena.Data$yPosition)
    Sys.sleep(.09)
  }
  return(gsiLog)
}

finalGsi <- time.stepper(10)

plot(finalGsi)

File: reynoldsSim.js
// # THE FOLLOWING IS A 2-D VARIANT OF CRAIG REYNOLDS' BOID MODEL.
// # Written by Gabriel Dell'Accio
// # Algorithm inspired by Craig Reynold's BOID algorithm and Conrad Parker's BOIDs pseudocode at: 
// # http://www.kfish.org/boids/pseudocode.html

// # Overview:
// #
// #   Rules for swarming
// #     1) Center of Mass: Boids are attracted to the center of the overall swarm.
// #     2) Avoidance rule: Boids avoid collision with other boids by avoiding coming within
// #        a certain distance of one another.
// #     3) Match Velocity: Boids can sense the average velocity of the swarm and do their 
// #        best to match it.
// # 
// #    GSI: Group Stability Index 
// #     # This is an index developed by Baldessare and colleagues (2003) to measure the stability of a 
// #     # swarm in their paper "Evolving mobile robots  able to display collective behaviors"


// Starting Variables
var nIndividuals = 10
var baseVelocity = 10
var spatialDistribution = 100
var centerStart = 500
var nSteps = 20
var bodySize = 5

// Load the gaussian package in order to sample from normal 
// distributions.
var normalDist = require("gaussian")
var matterJS = require("matter-js")
var bodies = matterJS.Bodies

// Array to hold the BOID swarm
var swarmArray = []

// Object that represents a boid.
function boid(count, locx, locy, xvel, yvel) {
  this.index = count
  this.xPosition = locx
  this.yPosition = locy
  this.xvlc = xvel
  this.yvlc = yvel
  this.body = bodies.circle(locx, locy, bodySize, bodySize)
  
}

// Helpper function for printing the location of a swarm.
function printBoidLocation(boid) {
  console.log("Boid loc: " +[boid.xPosition, boid.yPosition])
}

// 
function initializeSwarm(swarm) {
  for(h = 0; h < nIndividuals; h ++) {

    var distribution = normalDist(centerStart, spatialDistribution);

    // Generate x,y coordinates from the random distribution we generated.
    var tempX = distribution.ppf(Math.random());
    var tempY = distribution.ppf(Math.random());

    var tempBoid = new boid(h, tempX, tempY, baseVelocity, baseVelocity);
    
    swarm.push(tempBoid);

    // printBoidLocation(tempBoid)
  } 
  // console.log("Initialized Swarm:" + swarm)
  return swarm
}


function computeCartDist(x1,x2,y1,y2) {
  var a = x1 - x2 
  var b = y1 - y2 
  var c = Math.sqrt(a*a + b*b)
  return c
}

function calculateGroupDist(swarm) {
  var groupedDistance = 0
  for(i=0; i<nIndividuals; i++) {
    for(j=0;j<nIndividuals; j++) {
      var tempDist = computeCartDist(swarm[i].xPosition,swarm[j].xPosition,swarm[i].yPosition, swarm[j].yPosition);
      // console.log(tempDist)
      groupedDistance = groupedDistance + tempDist;
    } 
  }
  // console.log("Grouped distance: "+groupedDistance)
  return groupedDistance;
}

function calculateGSI(groupedDistancePrev,groupedDistanceNext) {
  // console.log("groupedDistancePrev: "+ groupedDistancePrev)
  // console.log("groupedDistanceNext: "+ groupedDistanceNext)
  var finalGSI = 1 - (((Math.abs(groupedDistancePrev - groupedDistanceNext)) / 4 ) / (nIndividuals*(nIndividuals-1)/2))
  // console.log(finalGSI)

  // console.log(finalGSI)
  return finalGSI
}

// Find the center of the swarm by finding the average x,y of the swarm.
function centerOfMass(boid, swarm) {
  var totalY = 0
  var totalX = 0
  for(t=0;t<nIndividuals;t++){
    totalX = totalX + swarm[t].xPosition
    totalY = totalY + swarm[t].yPosition
  }
  var averageX = totalX / nIndividuals
  var averageY = totalY / nIndividuals




  var centerMass = [((averageX - boid.xPosition)/100), ((averageY - boid.yPosition)/100)]

  // console.log("Center mass rule is: " + centerMass)

  return centerMass
}

function avoidRule(boid, swarm) {
  
  var correctedCourse = [0,0]

  for(it=0;it<nIndividuals;it++) {

    if(boid.index == it) { 

    } else {

      if((Math.abs(swarm[it].xPosition - boid.xPosition) < 100) && (Math.abs(swarm[it].yPosition - boid.yPosition) < 100)) {
        var currentPosition = [swarm[it].xPosition, swarm[it].yPosition]
        // console.log("Current Position for avoid rule: " + currentPosition)
        // console.log("Current Position for the boid: " + [boid.xPosition, boid.yPosition])
        // console.log((currentPosition[0] - boid.xPosition))
        // console.log((currentPosition[1] - boid.yPosition))

        correctedCourse[0] = correctedCourse[0] - (currentPosition[0] - boid.xPosition)

        // console.log("Correction x: " + correctedCourse[0])
        correctedCourse[1] = correctedCourse[1] - (currentPosition[1] - boid.yPosition)

        // console.log("Correction y: " + correctedCourse[1])
      }
    }

  }
  

  correctedCourse = [correctedCourse[0]/nIndividuals, correctedCourse[1]/nIndividuals]
  // console.log("avoid rule: " + correctedCourse)
  return correctedCourse 
} 


function matchVelocity(cBoid, swarm) {
  var meanVelocityX = 0
  var meanVelocityY = 0

  for(m=0; m<nIndividuals;m++) {
    meanVelocityY = meanVelocityY + cBoid.yvlc
    meanVelocityX = meanVelocityX + cBoid.xvlc
  }
  meanVelocityX = meanVelocityX / nIndividuals
  meanVelocityY = meanVelocityY / nIndividuals

  var velocityCorrectionX = (meanVelocityX - cBoid.xvlc) / 8 
  var velocityCorrectionY = (meanVelocityY = cBoid.yvlc) / 8

  // console.log("velocityRule: " + [velocityCorrectionX, velocityCorrectionY])

  return [velocityCorrectionX,velocityCorrectionY]
}


function computeNextPos(cSwarm) {

  // console.log("This is the cswarm:")
  // console.log(cSwarm)

  var newSwarm = []

  for(k=0; k < nIndividuals; k++) {
    

    var tempBoid = new boid(cSwarm[k].index, cSwarm[k].xPosition, cSwarm[k].yPosition, cSwarm[k].xvlc, cSwarm[k].yvlc);

    // console.log("NEW BOID")
    // console.log(tempBoid)
    var centerMassRule = centerOfMass(tempBoid, cSwarm)
    var avoidanceRule = avoidRule(tempBoid, cSwarm)
    var velocityRule = matchVelocity(tempBoid, cSwarm)

    tempBoid.xvlc = tempBoid.xvlc + centerMassRule[0] + avoidanceRule[0] + velocityRule[0]
    tempBoid.yvlc = tempBoid.yvlc + centerMassRule[1] + avoidanceRule[1] + velocityRule[1] 

    var tempBoidPosX = tempBoid.xPosition + tempBoid.xvlc
    var tempBoidPosY = tempBoid.yPosition + tempBoid.yvlc

    
    tempBoid.xPosition = tempBoidPosX
    tempBoid.yPosition = tempBoidPosY
    // console.log("NEW BOID AFTER CHANGE")
    // console.log(tempBoid)
    // console.log("Current ind:")
    // console.log(tempBoid)
    newSwarm.push(tempBoid)
    // cSwarm[i] = tempBoid
    // console.log("current individual after being moved: " + [tempBoid.xPosition, tempBoid.yPosition])
  }
  // console.log("THis is the new swarm:") 
  // console.log(newSwarm)
  return newSwarm
}


function computeBehaviors(currentSwarm) {

    var prevDist = calculateGroupDist(currentSwarm)

    var nextSwarm = computeNextPos(currentSwarm)
    var nextDist = calculateGroupDist(nextSwarm)

    var temp = calculateGSI(prevDist,nextDist)

    return [temp, nextSwarm]
    // console.log(nextDist)
}

function step(swarmList) {
  



 
  // for(i=0;i<steps;i++) {

  
    computeResult = computeBehaviors(swarmList)

    // console.log(computeResult)

    // console.log("hello")


    // console.log("current gsi log: " + computeResult[0])


    var distLog = (computeResult[0])


    // console.log("swarmState: " + swarmState)

    swarmState = computeResult[1]
  // }
  // console.log(dis)
  return [distLog, swarmState]
}




var swarmArray = initializeSwarm(swarmArray)
// console.log(swarmArray)


function stepper(swarm) {
  
  var theSwarm = swarm

  var gsiLog = []
  var swarmStateLog = []

  for(p=0;p<nSteps;p++) {
    // console.log("step: "+p)
    var result = step(theSwarm)

    var theSwarm = result[1]
    // console.log("The swarm: " + swarm)
    gsiLog.push(result[0])
    swarmStateLog.push(theSwarm)
    // console.log("gsi: " + gsi)
  }


  return [gsiLog, swarmStateLog]
}





result = stepper(swarmArray)
console.log(result[0])

module.exports.experimentResult = result;
// console.log()
// console.log(result)
// console.log()

// console.log(GSIresult)


/////////////////